nav2_docking/opennav_docking/src/controller.cpp

```diff
@@ -55,16 +81,120 @@ Controller::Controller(const rclcpp_lifecycle::LifecycleNode::SharedPtr & node)
+}
+
+Controller::~Controller()
+{
+  control_law_.reset();
+  trajectory_pub_.reset();
+  collision_checker_.reset();
+  costmap_sub_.reset();
+  footprint_sub_.reset();
 }
 
 bool Controller::computeVelocityCommand(
-  const geometry_msgs::msg::Pose & pose, geometry_msgs::msg::Twist & cmd, bool backward)
+  const geometry_msgs::msg::Pose & pose, geometry_msgs::msg::Twist & cmd, bool is_docking,
+  bool backward)
 {
   std::lock_guard<std::mutex> lock(dynamic_params_lock_);
   cmd = control_law_->calculateRegularVelocity(pose, backward);
+  return isTrajectoryCollisionFree(pose, is_docking, backward);
+}
+
+bool Controller::isTrajectoryCollisionFree(
+  const geometry_msgs::msg::Pose & target_pose, bool is_docking, bool backward)
+{
+  // Visualization of the trajectory
+  nav_msgs::msg::Path trajectory;
+  trajectory.header.frame_id = base_frame_;
+  trajectory.header.stamp = clock_->now();
+
+  // First pose
+  geometry_msgs::msg::PoseStamped next_pose;
+  next_pose.header.frame_id = base_frame_;
+  trajectory.poses.push_back(next_pose);
+
+  // Get the transform from base_frame to fixed_frame
+  geometry_msgs::msg::TransformStamped base_to_fixed_transform;
+  try {
+    base_to_fixed_transform = tf2_buffer_->lookupTransform(
+      fixed_frame_, base_frame_, trajectory.header.stamp,
+      tf2::durationFromSec(transform_tolerance_));
+  } catch (tf2::TransformException & ex) {
+    RCLCPP_ERROR(
+      logger_, "Could not get transform from %s to %s: %s",
+      base_frame_.c_str(), fixed_frame_.c_str(), ex.what());
+    return false;
+  }
+
+  // Generate path
+  for (double t = 0; t < projection_time_; t += simulation_time_step_) {
+    // Apply velocities to calculate next pose
+    next_pose.pose = control_law_->calculateNextPose(
+      simulation_time_step_, target_pose, next_pose.pose, backward);
+
+    // Add the pose to the trajectory for visualization
+    trajectory.poses.push_back(next_pose);
+
+    // Transform pose from base_frame into fixed_frame
+    geometry_msgs::msg::PoseStamped local_pose = next_pose;
+    local_pose.header.stamp = trajectory.header.stamp;
+    tf2::doTransform(local_pose, local_pose, base_to_fixed_transform);
+
+    // Determine the distance at which to check for collisions
+    // Skip the final segment of the trajectory for docking
+    // and the initial segment for undocking
+    // This avoids false positives when the robot is at the dock
+    double dock_collision_distance = is_docking ?
+      nav2_util::geometry_utils::euclidean_distance(target_pose, next_pose.pose) :
+      std::hypot(next_pose.pose.position.x, next_pose.pose.position.y);
+
+    // If this distance is greater than the collision_tolerance, check for collisions
+    auto projected_pose = nav_2d_utils::poseToPose2D(local_pose.pose);
+    if (use_collision_detection_ &&
+      dock_collision_distance > collision_tolerance_ &&
+      !collision_checker_->isCollisionFree(projected_pose))
```

Comment:
```suggestion
      !collision_checker_->isCollisionFree(nav_2d_utils::poseToPose2D(local_pose.pose)))
```

then you can remove L169

Super nit pick, only since you're going to go back in to update the tests